pilot/pkg/config/kube/crdclient/client.go

```diff
@@ -423,3 +459,56 @@ func genPatchBytes(oldRes, modRes runtime.Object, patchType types.PatchType) ([]
 		return nil, fmt.Errorf("unsupported patch type: %v. must be one of JSONPatchType or MergePatchType", patchType)
 	}
 }
+
+func handleCRDAdd(cl *Client, name string, stop <-chan struct{}) {
+	scope.Debugf("adding CRD %q", name)
+	s, f := cl.schemasByCRDName[name]
+	if !f {
+		scope.Debugf("added resource that we are not watching: %v", name)
+		return
+	}
+	resourceGVK := s.Resource().GroupVersionKind()
+	gvr := s.Resource().GroupVersionResource()
+
+	cl.kindsMu.Lock()
+	defer cl.kindsMu.Unlock()
+	if ch, f := cl.kinds[s.Resource().GroupVersionKind()]; f {
+		scope.Debugf("added resource that already exists: %v", resourceGVK)
+		// Wait for it to sync again. This ensures that SyncAll will properly wait
+		cache.WaitForCacheSync(nil, func() bool {
+			if ch.informer.HasSynced() {
+				return true
+			}
+			scope.Infof("waiting for %v to sync...", resourceGVK)
+			return false
+		})
+		return
+	}
+	var i informers.GenericInformer
+	var ifactory starter
+	var err error
+	if s.Resource().Group() == gvk.ServiceApisGateway.Group {
+		ifactory = cl.client.GatewayAPIInformer()
+		i, err = cl.client.GatewayAPIInformer().ForResource(gvr)
+	} else {
+		ifactory = cl.client.IstioInformer()
+		i, err = cl.client.IstioInformer().ForResource(gvr)
+	}
+
+	if err != nil {
+		// Shouldn't happen
+		scope.Errorf("failed to create informer for %v", resourceGVK)
+		return
+	}
+	cl.kinds[s.Resource().GroupVersionKind()] = createCacheHandler(cl, s, i)
+	if stop != nil {
+		// Start informer factory, only if stop is defined. In startup case, we will not start here as
+		// we will start all factories once we are ready to initialize.
+		// For dynamically added CRDs, we need to start immediately though
+		ifactory.Start(stop)
```

Comment:
looks there can start multi informer for a single crd, if it is created before istiod start

Reply:
the Start code does nothing if called multiple times , so it shouldn't be an issue. It's really "start all unstarted informers"